### ISSUE TITLE
403 when API issues redirect

### ISSUE THREAD

**Issue by elreydetodo** [akamai-open/AkamaiOPEN-edgegrid-java#21]
The library needs to re-sign requests when an API issues a redirect. This issue was reported in [this Akamai Community post](https://community.akamai.com/thread/4516).

**Comment by mgawinec** [comment]
I have reproduced the case by adding redirect tests for REST-assured and Google HTTP client in the linked branch commit.

There problem is with REST-assured: `com.akamai.edgegrid.signer.restassured.RestAssuredEdgeGridFilterTest#signFollowedRedirects` fails.

There's no way to solve the issue with current architecture of REST-assured EdgeGrid client, because redirection is handled on Apache HTTP Client level, while Signing Filter works above this level. The solution would be to release a completely different version of REST-assured EdgeGrid client that uses Apache HTTP Client request interceptors. IMHO it may help solving similar low-level issues in the future.

@elreydetodo Let me know how do you want to proceed.

**Comment by elreydetodo** [comment]
It looks like both the Google and REST Assured clients can use the Apache HTTP Client as their backing implementation to make the requests. We should add another module for it with an implementation of [RedirectStrategy](https://hc.apache.org/httpcomponents-client-4.5.x/httpclient/apidocs/org/apache/http/client/RedirectStrategy.html).

Google HTTP Client doesn't have its own sort of redirect handling system, just a basic config for whether to follow them or not. Maybe we could add a static method that takes an `ApacheHttpClient` and configures it with a `RedirectStrategy`. I don't know what we'd do for `NetHttpClient`. #17 calls for adding bindings for the JDK client, we may have to address that in order to fully address this issue.

REST Assured seems to expose the `HttpClient` in various places, but I don't know how it would fit into the normal use case to intercept that thing and attach a `RedirectStrategy` to it. On the plus side it looks to me like it _only_ supports that one client library, so we only need one sort of fix for it.

**Comment by mgawinec** [comment]
Google HTTP Client handles signing for redirect correctly (at list with Apache HTTP client, I haven't check it with other clients) so we don't need to handle that. My [test here][1] proves that: request interceptor signs not only original request but also followed redirect request. So, I wouldn't manipulate Google HTTP Client at all.

However, I like the idea of releasing a new module module for Apache HTTP Client only. It would offer `EdgeGridSigningInterceptor` (implementation of [`HttpRequestInterceptor`][2]) that is able to sign both original request and followed redirects. Then we would provide instruction how this can be used individually or in combination with REST-assured:

```java
RestAssured.given().config(RestAssuredConfig.config().httpClient(HttpClientConfig
                .httpClientConfig()
                .httpClientFactory(new
                   HttpClientConfig.HttpClientFactory() {
                       @Override
                       public HttpClient createHttpClient() {
                           AbstractHttpClient client = new DefaultHttpClient();
                           client.addRequestInterceptor(new EdgeGridSigningInterceptor());
                           return client;
                       }
                   })))
                   ...
```

That would, obviously, limit usage of "edgegrid-signer-rest-assured" to only OPEN APIs that do not have redirects. And if a user has an OPEN API with redirects, she would need to use our new Apache module.

If you see an advantage of using `RedirectStrategy` instead of that solution, let me know.

[1]: ttps://github.com/akamai-open/AkamaiOPEN-edgegrid-java/commit/4d0eea5e584c1ccf6497cee29a40e1f69608eb5f#diff-8e7aba6544d7dd034c24cc2e71e6a9e7
[2]: https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpRequestInterceptor.html

**Comment by elreydetodo** [comment]
Alright, I'm read more into the specific methods offered on the objects in the REST Assured library and I think the code printed above is correct. I think we should suggest that as an alternative to the default (the Filter implementation we currently have) in the event their API usage involves redirects.

It looks like `RedirectStrategy` wouldn't easily solve this problem, but we should write it anyway for anyone who is using `HttpClient` directly. I noticed that there's actually a method `HttpClientConfig#reuseHttpClientInstance()` which notes that "By default REST Assured will create a new HttpClient instance for each request." I would guess that a redirect constitutes a new request, so the `RedirectStrategy` might not even ever get called.

**Comment by elreydetodo** [comment]
The new Apache HTTP Client module has been merged and will be included in the 2.1.0 release. That should be released tomorrow once we've finished testing on it. There's also some pending changes to the README to make it easier to find relevant instructions about this problem.
